### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by Dobatymo**
Hi, this pull request would enable Windows support.

The following 3 things were necessary to make it work:

- remove the file locking as `fcntl` does not exist on Windows
- close the temporary file prior to removal as Windows would complain otherwise
- replace more characters in filenames like `<>:"/\\|?*` and ASCII control chars

The other changes are minor and not Windows specific.

### CODE REVIEW

**Comment by fnep** [comment]
In general, I would welcome windows support, but honestly your pull requests changes a lot more than needed for windows support. 

I'm not so happy with all the changes, so I will add some comments to the diff to discuss this further.

File: mtv_dl.py [hunk 1]
```diff
@@ -170,8 +168,12 @@
 from tqdm import tqdm
 from yaml.error import YAMLError
 
+system = platform.system()
+
 CHUNK_SIZE = 128 * 1024
 
+LIMIT_DESC = 40
```
Comment (fnep) [inline mtv_dl.py]: This is not related to windows. If you want such a feature, put it in a separate request.

File: mtv_dl.py [hunk 2]
```diff
@@ -170,8 +168,12 @@
 from tqdm import tqdm
 from yaml.error import YAMLError
 
+system = platform.system()
```
Comment (fnep) [inline mtv_dl.py]: This variable is only used once. Please use platform.system() directly.

File: mtv_dl.py [hunk 3]
```diff
@@ -145,17 +145,15 @@
 import urllib.parse
 from pathlib import Path
 from contextlib import contextmanager
-from datetime import datetime, timedelta
+from datetime import datetime, timedelta, timezone
 from functools import partial
-from itertools import chain
-from itertools import islice
+from itertools import chain, islice
```
Comment (fnep) [inline mtv_dl.py]: Please import only one module per line (same for timezone). This to simplify merges.

File: mtv_dl.py [hunk 4]
```diff
@@ -306,13 +324,13 @@ def wrapped(*args, **kwargs):
             try:
                 with _cache_file.open('rb') as cache_handle:
                     with flocked(cache_handle, timeout=60):
-                        logger.debug("Using cache from %r.", _cache_file)
+                        logger.debug("Using cache from %s.", _cache_file)
                         return pickle.load(cache_handle)
-            except Exception:
+            except Exception as e:
```
Comment (fnep) [inline mtv_dl.py]: `e` is unsued.

File: mtv_dl.py [hunk 5]
```diff
@@ -306,13 +324,13 @@ def wrapped(*args, **kwargs):
...
                 res = fn(*args, **kwargs)
                 with _cache_file.open('wb') as cache_handle:
                     with flocked(cache_handle, timeout=60):
-                        logger.debug("Saving cache to %r.", _cache_file)
+                        logger.debug("Saving cache to %s.", _cache_file)
```
Comment (fnep) [inline mtv_dl.py]: Why not `%r`? I like to see the repr in the logs.

File: mtv_dl.py [hunk 6]
```diff
@@ -306,13 +324,13 @@ def wrapped(*args, **kwargs):
             try:
                 with _cache_file.open('rb') as cache_handle:
                     with flocked(cache_handle, timeout=60):
-                        logger.debug("Using cache from %r.", _cache_file)
+                        logger.debug("Using cache from %s.", _cache_file)
```
Comment (fnep) [inline mtv_dl.py]: Why not `%r`?

File: mtv_dl.py [hunk 7]
```diff
@@ -368,13 +386,13 @@ def _showlist(self, retries: int = len(DATABASE_URLS)) -> Generator[Path, None,
             retries -= 1
             try:
                 url = random.choice(DATABASE_URLS)
-                logger.debug('Opening database from %r.', url)
+                logger.debug('Opening database from %s.', url)
```
Comment (fnep) [inline mtv_dl.py]: `%r`

File: mtv_dl.py [hunk 8]
```diff
@@ -268,27 +270,43 @@ def serialize_for_json(obj: Any) -> str:
...
+                    break
 
-    finally:
-        fcntl.flock(fd, fcntl.LOCK_UN)
+        finally:
+            fcntl.flock(fd, fcntl.LOCK_UN)
+
+    invalid_chars = '/\0'
+
+invalidre = re.compile("[{}]".format(re.escape(invalid_chars)))
```
Comment (fnep) [inline mtv_dl.py]: To make clear this is a constant, please use uppercase and underscore naming.

I also see no good reason to use re.escape here: Better the compiled regex in the if-else above.

File: mtv_dl.py [hunk 9]
```diff
@@ -410,11 +429,10 @@ def _load(self) -> Dict[str, Union[float, List, Dict[str, Any]]]:
         channel, topic, region = None, None, None
 
         with self._showlist() as showlist_path:
-            with lzma.open(showlist_path) as fh:
+            with lzma.open(showlist_path, 'rt', encoding='utf-8') as fh:
```
Comment (fnep) [inline mtv_dl.py:432]: This is a nice change. I did not know it existed.

File: mtv_dl.py [hunk 10]
```diff
@@ -850,7 +868,7 @@ def download(self,
                        or self["url_http%s" % quality[1]] \
                        or self["url_http%s" % quality[2]]
 
-            logger.debug('Downloading %s from %r.', self.label, show_url)
+            logger.debug('Downloading %s from %s.', self.label, show_url)
```
Comment (fnep) [inline mtv_dl.py]: `%r` please.

File: mtv_dl.py [hunk 11]
```diff
@@ -872,15 +890,15 @@ def download(self,
                 return None
 
             if include_subtitles and self['url_subtitles']:
-                logger.debug('Downloading subtitles for %s from %r.', self.label, self['url_subtitles'])
+                logger.debug('Downloading subtitles for %s from %s.', self.label, self['url_subtitles'])
```
Comment (fnep) [inline mtv_dl.py]: `%r` please.

File: mtv_dl.py [hunk 12]
```diff
@@ -715,21 +733,21 @@ def _move_to_user_target(self,
...
+                                                              time=self['start'].strftime('%H-%M'),
                                                               **escaped_show_details))
 
         destination_file_path.parent.mkdir(parents=True, exist_ok=True)
         try:
-            shutil.move(source_path.as_posix(), destination_file_path.as_posix())
+            shutil.move(source_path, destination_file_path)
         except OSError as e:
-            logger.warning('Skipped %s: %s', self.label, str(e))
+            logger.warning('Skipped %s. Moving %s to %s failed: %s', self.label, source_path, destination_file_path, e)
```
Comment (fnep) [inline mtv_dl.py]: `%r` please.

File: mtv_dl.py [hunk 13]
```diff
@@ -487,7 +505,7 @@ def clear(self):
         except OSError:
             pass
         else:
-            logger.debug("Dropped cache %r.", self.cache_file_path)
+            logger.debug("Dropped cache %s.", self.cache_file_path)
```
Comment (fnep) [inline mtv_dl.py]: `%r` please.

**Review by fnep — CHANGES_REQUESTED** [review]
My comments seem to be very pedantic, so please don't get me wrong. I like the change, but I think here is a lot about personal preferences and about the actual change.

**Comment by Dobatymo** [comment]
Thanks for your feedback. I should have limited the PR to Windows support in the first place. I think I have fixed all your comments except the one regarding `re.escape`. Why is it not necessary?

While debugging I was annoyed by all the WindowsPath(asd/asd) stuff instead of only 'asd/asd'. That's why I left the `%s` in the code before.

I squashed the commits to keep the history clean.

**Comment by fnep** [comment]
I had another look and think its fine now.  👍

Thanks a lot for you work!

--- 

I have no time these days to publish it as a new release, but i will do so as soon as possible. I also thought about using sqlite instead of pickle to cache the Filmliste file. This way we would also avoid the locking issues in future and propably its even faster. But i need a little bit mor time to test this idea.

**Pull request merged**